### ISSUE TITLE
Update macOS SDK

### ISSUE THREAD

**Issue by r-burns** [NixOS/nixpkgs#101229]
I'm seeing a growing number of failures on hydra for packages requiring the 10.13 SDK on macOS. What is the timescale on a possible update, and what would that involve? Alternatively, would there be some way to build certain packages against the 10.13 SDK while keeping the default stdenv and majority of packages 10.12-compatible?

Related:
https://github.com/NixOS/nixpkgs/pull/99133 (folly)
https://github.com/NixOS/nixpkgs/pull/101125 (apk-tools)
https://github.com/NixOS/nixpkgs/issues/97479 (various)

**Comment by prusnak** [comment]
According to https://en.wikipedia.org/wiki/MacOS_version_history#Releases

* macOS 10.12 was EOL on October 2019 
* macOS 10.13 will be EOL on November 30, 2020

So I guess ditching 10.12 is fine

**Comment by siraben** [comment]
In general how long are we comfortable using EOL'd SDKs? Are there security implications?

**Comment by siraben** [comment]
I located the URL for the 10.13 SDK, http://swcdn.apple.com/content/downloads/50/15/041-91747-A_WICZE7RNVZ/rayjnqf847xflt3tan8o8agod67eq88cav/DevSDK_macOS1013_Public.pkg

**Comment by abathur** [comment]
@LnL7 @thefloweringash Not sure who all should be in the loop here, but I saw a notification about this.

**Comment by jonringer** [comment]
cc @NixOS/darwin-maintainers

**Comment by LnL7** [comment]
Yeah, this is probably something that should be tackled (along with reviving the llvm update) once Big Sur support is in.

**Comment by SuperSandro2000** [comment]
I just encountered a python module which requires [SystemExtensions](https://developer.apple.com/documentation/systemextensions) which is only available on 10.15+.

**Comment by SuperSandro2000** [comment]
I created #101940 to resolve this Issue. I did so far no real testing and would appreciate some help and advice.

**Comment by SuperSandro2000** [comment]
I also found https://github.com/phracker/MacOSX-SDKs. I am not sure at all if this is a trust worthy source but maybe for testing a good option.

**Comment by siraben** [comment]
10.13 is being EOL'd soon, as during the last decade Apple generally ends support for an OS (and thus its SDK) three years later. https://en.wikipedia.org/wiki/MacOS_version_history#Releases . Should we update to 10.14 SDK instead?

**Comment by SuperSandro2000** [comment]
I did not find a link (https://github.com/NixOS/nixpkgs/pull/101940/files#diff-c4d8bbc3261c329aad6d37bf66765868a01dcf4b838d8622d16959d8597ebed8R20) for 10.14. Maybe someone else has better luck than me.

**Comment by LnL7** [comment]
I'm not sure what sdk this is we're using right now but the CLT is available there which should also contain everything we need.

```
http://swcdn.apple.com/content/downloads/41/57/061-26573-A_JMOA8GZGDR/lj8yrtu8dgs40fw9k8f5fkoviwkp0og6vs/CLTools_SDK_macOS1014.pkg
```

**Comment by siraben** [comment]
Related:
#106679 (fails to compile because requires newer SDK)

**Comment by prusnak** [comment]
Maybe we should update to SDK 10.14, not 10.13, because https://github.com/NixOS/nixpkgs/issues/95199 mentions that Qt 5.14+ requires 10.14 SDK

**Comment by SuperSandro2000** [comment]
@prusnak I could not find the link for 10.14.

**Comment by r-burns** [comment]
I think based on apple's previous EOL schedules, they will probably be dropping support for 10.13 in the next few months now that big sur is out, so jumping to 10.14 could be timely

**Comment by siraben** [comment]
Agreed, going to 10.14 is better because 10.13 will be EOL'd soon.

**Comment by holymonson** [comment]
I'm building webkitgtk and find it requires 4 backport of macos system headers. (Related #106818 #107189)
```
    # FIXME: remove below patch after libproc.h update to 10.12 or above
    ./fix-WTF-headers.patch
    # FIXME: remove below patch after Libsystem-osx update to 10.14 or above
    ./fix-CommonCrypto-headers.patch
    # FIXME: remove below patch after Libsystem-osx update to 10.15 or above
    ./fix-qos-headers.patch
    # FIXME: remove below patch after Libsystem-osx update to 11.1 or above
    ./backport-MSG_NOSIGNAL-in-socket_h.patch
```

I'm surprised that we are stuck on an outdated SDK so long (FYI, homebrew will drop support on EOL versions immediately). Since Nix is so good at version management and overwriting, why don't we offer higher SDKs that some packages could depend on? (Of course, base packages should always depend on the lowest one). So here is several choices:
1. Update to SDK 10.14, and do nothing on higher SDK. For packages requiring higher SDKs, backport on their own.
2. Provide several versions of higher SDK that packages could explicitly demand. We may need a variable like `NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM` to allow low macOS to use high SDK.
3. Split the whole SDK into independent libraries (and keep a lowest SDK), so packages could demand their necessary ones.

I know maintainers are busy dealing with Apple Silicon, but this issue deserves more attention. Thanks.

**Comment by SuperSandro2000** [comment]
> why don't we offer higher SDKs that some packages could depend on?

There is no reason to maintain two or more versions of the SDK. This would also mean we would need to maintain two versions of every tool listed here https://opensource.apple.com/release/macos-10136.html.

The update is simple blocked by lack of time. Someone just needs to do it which makes this entire discussion a bit pointless.

**Comment by FRidh** [comment]
Mesa in staging-next requires 10.13+. https://github.com/NixOS/nixpkgs/pull/107783#issuecomment-752415481

**Comment by siraben** [comment]
Is there a certain threshold until it becomes imperative to update to 10.13? I'd like to help efforts as well but there's a lot to do and it would be nice if we had a tracking project.

**Comment by SuperSandro2000** [comment]
I would do it but stdenv dies currently not build in Darwin which is already a bad starting point.

**Comment by FRidh** [comment]
Qt 5.14 is EOL and will be removed. Packages should be migrated to 5.15, however, its not supported with the current SDK. https://github.com/NixOS/nixpkgs/pull/108129

**Comment by prusnak** [comment]
> Qt 5.14 is EOL and will be removed. Packages should be migrated to 5.15, however, its not supported with the current SDK. #108129

macOS build uses Qt 5.12 which is a LTS release (supported until December 2021).

**Comment by Takangcwele** [comment]
#113290

**Comment by r-burns** [comment]
From the Go 1.16 release notes:

>  Go 1.16 is the last release that will run on macOS 10.12 Sierra. Go 1.17 will require macOS 10.13 High Sierra or later.

https://golang.org/doc/go1.16#darwin

**Comment by uri-canva** [comment]
> I would do it but stdenv does currently not build in Darwin which is already a bad starting point.

@SuperSandro2000 Does https://github.com/NixOS/nixpkgs/pull/107013 address that or are stdenv and bootstrap stdenv two different things, and fixing bootstrap stdenv isn't enough to fix stdenv?

**Comment by SuperSandro2000** [comment]
> > I would do it but stdenv does currently not build in Darwin which is already a bad starting point.
> 
> @SuperSandro2000 Does https://github.com/NixOS/nixpkgs/pull/107013 address that or are stdenv and bootstrap stdenv two different things, and fixing bootstrap stdenv isn't enough to fix stdenv?

I need to correct myself a bit: it does not build for me but my darwin machine is a bit special and couldn't build it so far for reasons i do not know.

Bootstrapping stdenv is required to build a clean stdenv we can use to build other things. Stdenv can be broken without bootstraping being broken but not the other way around but I do not know if it is broken right now or works.

**Comment by TerrorJack** [comment]
The link for 10.15 SDK is http://swcdn.apple.com/content/downloads/45/28/061-26566-A_R7LP2VO5QF/svs27at6zihaghh3wj0av4y3wdcriu4n2j/CLTools_macOS1015_SDK.pkg, discovered from https://swscan.apple.com/content/catalogs/others/index-10.15.merged-1.sucatalog, in case it helps.

**Comment by danderson** [comment]
Came here after trying to upgrade Tailscale to 1.16.0, and discovering that it's impossible because it requires Go >=1.17, buildGo117Module is disabled, and I _think_ this is the remaining blocker for reenabling it?

It's not clear to me from the issue history: is someone working on this? If not, what's the next step? I don't macOS much, but if someone has some general directions I'm willing to try.

**Comment by abathur** [comment]
@toonn has been working toward this, though [the latest update](https://discourse.nixos.org/t/nix-macos-monthly/12330/) indicated that this was a little gummed up around some missing bits. I don't have much sense of the practical implications of that, or if there are obvious places where hands might help.

**Comment by domenkozar** [comment]
@danderson We're currently in an unfortunate position, Apple doesn't publish all sources to be able to compile SDK anymore, nor do they publish any documentation.

@toonn has been working on reverse-engineering the build process and it's a slow process, so he'll opt for binary blobs to unblock this work and speed up the SDK bump.

See https://opencollective.com/nix-macos/updates for the funding campaign of this work.

**Comment by jonringer** [comment]
I'm going to remove this from 21.11 blockers, as the window has passed

**Comment by toonn** [comment]
Yeah, no problem, this depends on the LLVM 11 bump anyway, so there's no way to get it in.

**Comment by prusnak** [comment]
It seems Apple has revamped their Open Source website which contains more info than it used to: https://opensource.apple.com/releases/

Also source code links now point to a new GitHub organization: https://github.com/apple-oss-distributions

**Comment by toonn** [comment]
Interesting, I can't find the "more info" though. Do you mean the Projects page?

(I am noticing the releases page loads much slower though : / )

**Comment by domenkozar** [comment]
@toonn can you give a short update on SDK bumping? 

The good news is that we've growing the budget at https://opencollective.com/nix-macos so Toonn can spend more time on this in the following months.

Also if someone wants to be paid for this tedious work, let me know and you can help out Toonn.

**Comment by toonn** [comment]
I'm currently stuck on a Python issue.

```
Building setuptools wheel...
/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/setuptools /priva
Traceback (most recent call last):
  File "/nix/store/fkvh5szkvi5c86936p58gk9v0nhlq5gd-python3-3.9.10/lib/python3.9
    return _run_code(code, main_globals, None,
  File "/nix/store/fkvh5szkvi5c86936p58gk9v0nhlq5gd-python3-3.9.10/lib/python3.9
    exec(code, run_globals)
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from pip._internal.cli.main import main as _main
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from pip._internal.cli.autocompletion import autocomplete
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from pip._internal.cli.main_parser import create_main_parser
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from pip._internal.cli import cmdoptions
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from pip._internal.cli.parser import ConfigOptionParser
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from pip._internal.configuration import Configuration, ConfigurationError
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from pip._internal.exceptions import (
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from pip._vendor.requests.models import Request, Response
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from . import utils
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from ._internal_utils import to_native_string
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from .compat import is_py2, builtin_str, str
  File "/private/tmp/nix-build-python3.9-bootstrapped-pip-21.3.1.drv-0/pip/src/p
    from urllib.request import parse_http_list, getproxies, proxy_bypass, proxy_
  File "/nix/store/fkvh5szkvi5c86936p58gk9v0nhlq5gd-python3-3.9.10/lib/python3.9
    from _scproxy import _get_proxy_settings, _get_proxies
ModuleNotFoundError: No module named '_scproxy'
```

This started happening when I replaced configd from the source releases with SystemConfiguration from the SDK (because the configd releases from Apple no longer build).

I have yet to figure out why urllib is trying to import `_scproxy` now. Commit 6328518e removed this line:

```
-    substituteInPlace Lib/urllib.py --replace "if sys.platform == 'darwin'" "if False"
```

That was preventing the `_scproxy` import but hasn't been part of the CPython expression for 2 years. So this breaking now is surprising.

Bumping configd is turning out to be particularly challenging. I don't expect other source releases to be as hard to update but this makes it difficult to put an ETA on things.

I have considered simply dropping the source releases altogether, as was done for aarch64-darwin but this didn't really make things easier.

**Comment by willcohen** [comment]
As noted a few times in various issues, `go` now requires `_utimensat`, which is 10.13+ only, so we'll likely see version bumps on go projects increasingly failing to work on x86 darwin going forward.

**Comment by toonn** [comment]
I've finally managed to substitute SystemConfiguration for configd so source release bumping is currently progressing. This won't make the current release but I'm confident the bump'll be ready by next one.

**Comment by zowoq** [comment]
> `go` now requires `_utimensat`, which is 10.13+ only

`go` also wants `_SecTrustEvaluateWithError` which is 10.14+, some `rust` packages will want it as well.

**Comment by r-burns** [comment]
See also https://github.com/NixOS/nixpkgs/pull/144644, sometimes Go rolls its own toolchain so we don't need to worry about the version in stdenv at least for Go packaging purposes.

**Comment by zowoq** [comment]
> See also #144644, sometimes Go rolls its own toolchain so we don't need to worry about the version in stdenv at least for Go packaging purposes.

We got lucky with `go` for 1.17, despite what the upstream docs said we were mostly fine with it on the 10.12 SDK. Not so much with 1.18, we already have plenty of broken `go` packages.

**Comment by dhess** [comment]
To be clear, is the `go` breakage only affecting `x86_64-darwin` `nixpkgs`, or is it also affecting `aarch64-darwin`?

**Comment by zowoq** [comment]
> To be clear, is the `go` breakage only affecting `x86_64-darwin` `nixpkgs`, or is it also affecting `aarch64-darwin`?

`x86_64-darwin` only.

**Comment by toonn** [comment]
@burdiyan, Randy's done some work in #176661 to make it possible to use the 11.0.1 SDK on x86_64-darwin too. Longer-term the plan is to provide all recent SDKs in a similar fashion. The idea would be that we can build packages with a newer SDK if there's no alternative solution.

Maybe this can help you smooth out the differences between the platforms?

**Comment by reckenrode** [comment]
I opened #180931 to update things that were marked broken on x86_64-darwin. I’ll also look over this issue for things I can add to it. There are a handful of things out of scope for that PR.

Based on feedback from @toonn, I’ve reworked #176661 to bootstrap an 11.0 stdenv for x86_64-darwin. I ran into a blocker yesterday, but I think I got it solved. Unfortunately, it’s going to require a new bootstrap-tools for x86_64-darwin, and I’m not sure how exactly to make that happen.

**Comment by thiagokokada** [comment]
Sorry, not sure how merging my PR closed this issue (I didn't use any of the keywords that should have made this).

**Comment by mweinelt** [comment]
> Sorry, not sure how merging my PR closed this issue (I didn't use any of the keywords that should have made this).

Fixed: #issue

Is what probably caused this.

**Comment by thiagokokada** [comment]
> > Sorry, not sure how merging my PR closed this issue (I didn't use any of the keywords that should have made this).
> 
> Fixed: #issue
> 
> Is what probably caused this.

Yeah, I know about this keyword, but I double checked and I didn't use it at the PR.

**Comment by mweinelt** [comment]
> So we probably need to wait until this issue is fixed: #101229

I think you did.

**Comment by uri-canva** [comment]
Do we want to push the use of `apple_sdk_11_0` more broadly or should we only use it when necessary? I ask because some derivations might build with either one, if we want to flip the default stdenv to use `apple_sdk_11_0` eventually I'll start peppering `apple_sdk_11_0` as I happen upon them, but if `apple_sdk_11_0` is still a bit of a work in progress and we want to keep more flexibility in making changes to it in non-compatible ways I'll hold off.

**Comment by SuperSandro2000** [comment]
> should we only use it when necessary?

For now I would just keep doing that until takes overhand.

**Comment by mitchmindtree** [comment]
Is anyone successfully using the 10.13 SDK in their Nix darwin builds somehow?

Curious if there is some external repo hosting all known SDKs, even if they're not available in nixpkgs yet?

I think we'll likely start to see more issues around this now that GitHub [have updated their `macos-latest` runner to 10.13](https://github.blog/changelog/2023-04-24-github-actions-macos-13-is-now-available/) - as a result, some libraries that conditionally target newer APIs by checking the underlying OS version at build time are going to start failing with linker errors when they try to link to SDK functions that only exist in 10.13 as nixpkgs is only providing the 10.12 SDK.

We've run into this recently attempting to build the `sysinfo` crate depenency on the new `macos-latestl` runner https://github.com/FuelLabs/fuel.nix/issues/64 - it appears to attempt to use [the `_kCFURLVolumeAvailableCapacityForImportantUsageKey` API](https://developer.apple.com/documentation/corefoundation/kcfurlvolumeavailablecapacityforimportantusagekey) when built on 10.13 which is only available on 10.13+.

Are there any active WIPs adding `apple_sdk_13_0` to nixpkgs?

**Comment by mitchmindtree** [comment]
Ahh I'm mistaken - it looks like we're still encountering the same issue with the macOS runner on `macos-12`, and that both `macos-latest` and `macos-12` are still using `macos 12.6.5` anyway!

It seems the source of non-determinism in our macos builds must be coming from elsewhere. 

Apologies for the noise!

**Comment by reckenrode** [comment]
> We've run into this recently attempting to build the sysinfo crate depenency on the new macos-latest runner https://github.com/FuelLabs/fuel.nix/issues/64 - it appears to attempt to use [the _kCFURLVolumeAvailableCapacityForImportantUsageKey API](https://developer.apple.com/documentation/corefoundation/kcfurlvolumeavailablecapacityforimportantusagekey) when built on 10.13 which is only available on 10.13+.

I wanted to address this separately. Depending on the 11.0 SDK should be sufficient. According to the Apple documentation, that API was added in 10.13 (which is not the same thing as macOS 13).

**Comment by reckenrode** [comment]
> Apologies for the noise!

It was a good prompt to link some of the recent work on SDK- and stdenv-related stuff. There’s still work to be done after that on new SDK patterns to allow incremental “source-ification” of these SDKs, but it’ll allow more things to be built from source on a modern SDK (e.g., MoltenVK should be built against latest, Blender requires 12.x, etc).

**Comment by reckenrode** [comment]
I am working on refactoring Darwin and how SDKs are used. The new pattern constructs a full SDK, which can be added as a package to `buildInputs` to change the SDK used in the stdenv. No more `callPackage`. No more `overrideSDK`. `xcbuild` will also automatically use the SDK set up in the stdenv. It is no longer hardcoded. The following SDKs will be supported: 10.12.2, 10.13.2, 10.14.6, 10.15.6, 11.3, 12.3, 13.3, 14.4. The default SDKs are not changing, but changing them should be easy.

I have all of the above SDKs building some packages via an adapter. I have MoltenVK and Wine building against the 14.4 SDK. I have some source release packages (libutil) building with the 14.4 SDK and an older deployment target, so they are usable on old platforms still. I have PyTorch building from source with MPS support (addressing https://github.com/NixOS/nixpkgs/issues/243868). Once I have the bootstrap done, I can test wezterm (addressing https://github.com/NixOS/nixpkgs/issues/239384). So far, adding the new SDK has Just Worked™.

The timeline for this will depend on how long it takes me to get it bootstrapping. The bootstrap changes are non-trivial, but it appears I can do it without needing an update to the bootstrap tools. Thankfully, they include xar and cpio on both platforms. I also need libtapi and jq, but the question is whether it is possible to build them somehow first. In the absolute worst case, I can use a special extra bootstrap mode that doesn’t do the cleanups and processing required for a proper SDK in nixpkgs. I’d like to get this into the next staging-next cycle, but that could be aggressive. It’s definitely coming for 24.11.

**Comment by emilazy** [comment]
Linking https://github.com/NixOS/nixpkgs/pull/338695 here for those subscribed to this issue.

**Comment by reckenrode** [comment]
https://discourse.nixos.org/t/on-the-future-of-darwin-sdks-or-how-you-can-stop-worrying-and-put-the-sdk-in-build-inputs/50574 is the Discourse discussion with the plan for updating the SDKs and my current status.

**Comment by aviallon** [comment]
Closed by https://github.com/NixOS/nixpkgs/pull/346043
